lens/registry.go

```diff
@@ -0,0 +1,323 @@
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package lens
+
+import (
+	"context"
+	"encoding/json"
+
+	"github.com/ipfs/go-datastore/query"
+	"github.com/lens-vm/lens/host-go/config"
+	"github.com/sourcenetwork/immutable/enumerable"
+
+	"github.com/sourcenetwork/defradb/client"
+	"github.com/sourcenetwork/defradb/core"
+	"github.com/sourcenetwork/defradb/datastore"
+)
+
+// todo: This file, particularly the `lensLocker`` stuff, contains fairly sensitive code that is both
+// cumbersome to fully test with integration/benchmark tests, and can have a significant affect on
+// the users if broken (deadlocks, large performance degradation).  It should have proper unit tests.
+// https://github.com/sourcenetwork/defradb/issues/1596
+
+// lensRegistry is responsible for managing all migration related state within a local
+// database instance.
+type lensRegistry struct {
+	// lens lockers by source schema version ID
+	lensWarehouse map[string]*lensLocker
+	// lens configurations by source schema version ID
+	lensConfigs map[string]client.LensConfig
+}
+
+var _ client.LensRegistry = (*lensRegistry)(nil)
+
+func NewRegistry() *lensRegistry {
+	return &lensRegistry{
+		lensWarehouse: map[string]*lensLocker{},
+		lensConfigs:   map[string]client.LensConfig{},
+	}
+}
+
+func (r *lensRegistry) SetMigration(ctx context.Context, txn datastore.Txn, cfg client.LensConfig) error {
+	key := core.NewSchemaVersionMigrationKey(cfg.SourceSchemaVersionID)
+
+	json, err := json.Marshal(cfg)
+	if err != nil {
+		return err
+	}
+
+	err = txn.Systemstore().Put(ctx, key.ToDS(), json)
+	if err != nil {
+		return err
+	}
+
+	err = r.cacheLens(txn, cfg)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *lensRegistry) cacheLens(txn datastore.Txn, cfg client.LensConfig) error {
+	locker, ok := r.lensWarehouse[cfg.SourceSchemaVersionID]
+	if !ok {
+		locker = newLocker(cfg)
+		r.lensWarehouse[cfg.SourceSchemaVersionID] = locker
+	}
+
+	newLensPipes := make([]*lensPipe, LENS_POOL_SIZE)
+	for i := 0; i < LENS_POOL_SIZE; i++ {
+		var err error
+		newLensPipes[i], err = newLensPipe(cfg)
+		if err != nil {
+			return err
+		}
+	}
+
+	// todo - handling txns like this means that the migrations are not available within the current
+	// transaction if used for stuff (e.g. GQL requests) before commit.
+	// https://github.com/sourcenetwork/defradb/issues/1592
+	txn.OnSuccess(func() {
+	drainLoop:
+		for {
+			select {
+			case <-locker.safes:
+			default:
+				break drainLoop
+			}
+		}
+
+		for _, lensPipe := range newLensPipes {
+			locker.returnLens(lensPipe)
+		}
+
+		r.lensConfigs[cfg.SourceSchemaVersionID] = cfg
+	})
+
+	return nil
+}
+
+func (r *lensRegistry) RefreshLenses(ctx context.Context, txn datastore.Txn) error {
```

Comment:
note: This is tested, but only using badger-file, as the restart action is ignored for in-memory databases and code-cov is calculated using an in-memory store.